### ISSUE TITLE
FOUC with CSS Modules and Next.js 13.4.19

### ISSUE THREAD

**Issue by olalonde** [vercel/next.js#55451]
### Link to the code that reproduces this issue or a replay of the bug

https://github.com/olalonde/nextjs-fouc

### To Reproduce

https://github.com/olalonde/nextjs-fouc

npm install
npm start

Refresh localhost:3000 multiple times, you will see the content flash and move. 

### Current vs. Expected behavior

There should be no flash of unstyled content.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Mon Apr 24 20:53:44 PDT 2023; root:xnu-8796.121.2~5/RELEASE_ARM64_T8103
    Binaries:
      Node: 20.5.0
      npm: 9.8.0
      Yarn: 1.22.19
      pnpm: 7.30.0
    Relevant Packages:
      next: 13.4.20-canary.31
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.0.4
    Next.js Config:
      output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

**Comment by olalonde** [comment]
Oh wait, I believe it's fixed in canary. Hope it was an intentional fix :D Feel free to close.

**Comment by olalonde** [comment]
Nope, still seeing the FOUC in my other repo. I will try another reproduction...

**Comment by olalonde** [comment]
https://github.com/vercel/next.js/assets/416585/fec43aea-276e-48f9-b267-dc330e47d67c

**Comment by GabenGar** [comment]
Where do you get your styles? I don't see any `.css` imports inside [layout component](https://github.com/olalonde/nextjs-fouc2/blob/master/src/components/Layout/MainLayout.tsx). Also making the root layout a client component is a bad idea since it opts out of various optimizations and basically only renders after the script is parsed (which happens way after HTML and CSS) on client, like a typical client-only react app.

**Comment by olalonde** [comment]
Here: https://github.com/olalonde/nextjs-fouc2/blob/master/src/app/layout.tsx#L1

I was told I had to make my layout a client component by the Mantine people on Discord (apparently Mantine doesn't support server components).

**Comment by GabenGar** [comment]
Does it flash if you remove Mantine provider?

**Comment by olalonde** [comment]
Yes, it still flashes when I remove Mantine provider. See this branch: https://github.com/olalonde/nextjs-fouc2/tree/no-mantine-core

**Comment by kenjirohayashi** [comment]
I am facing the same behavior.

See this repository: https://github.com/kenjirohayashi/fouc-app-router

this is a plain repository created by `pnpm create next-app`

Refresh localhost:3000 several times, you'll see FOUC.



https://github.com/vercel/next.js/assets/79906992/3e02efbf-3f7c-414f-b9b3-a2b87ab5c684


### environment information
```
Relevant Packages
    react": 18.0.0
    next": 13.5.4
Binaries:
      pnpm: 8.1.0
```

**Comment by kenjirohayashi** [comment]
@pkasidit 
As you can see in my repository, I don't use "validatePath" in Server Action, but the FOUC occurs.
So your problem and mine may be due to different cause.
In my case, the FOUC probably occurs because judges that ["load" is completed](https://developer.chrome.com/docs/devtools/network/reference/#load)(red line in devtools waterfall) before "not-found.module.css" is load.

**Comment by claus** [comment]
I think this is due to how app router apps add page-level CSS to the `<head>`.

![Screenshot 2023-10-30 at 19 16 42](https://github.com/vercel/next.js/assets/2277/e6e35a79-bfff-4789-b90e-0e58371f86fc)

I hope i'm wrong or i'm missing something, because that would be very bad.

**Comment by calvinbrewer** [comment]
Experiencing the same issue with the following versions:

```
"next": "13.5.4",
 "react": "18.2.0",
 "react-dom": "18.2.0",
```

I'm seeing the same behavior as @claus reported. The browser starts rendering the document and requires an additional network request to fetch the CSS async causing FOUC. I'd assume this is to improve performance but at the cost of user experience. There should at least be a configuration to decide how you want styles loaded in the document.

![image](https://github.com/vercel/next.js/assets/10352306/ee1246c5-b3d5-4918-bb39-f6be715e517b)

**Comment by william-will-angi** [comment]
@GabenGar Are you saying that there is no way to preload css that isn't in your layout file?

**Comment by GabenGar** [comment]
Depends on what you meant by "preload".

**Comment by janrembold** [comment]
Are there any updates on that topic? I just installed Mantine in my project following the instructions for NextJS with App Router and still see a FOUC.
```
"next": "14.2.1"
"@mantine/core": "7.8.1"
```
